packages/x-data-grid/src/hooks/features/rows/useGridRowSpanning.ts

```diff
@@ -0,0 +1,293 @@
+import { useGridVisibleRows } from '../../utils/useGridVisibleRows';
+import { gridRenderContextSelector } from '../virtualization/gridVirtualizationSelectors';
+import { GridRenderContext } from '../../../models';
+import { useGridSelector } from '../../utils/useGridSelector';
+
+export interface GridRowSpanningState {
+  spannedCells: Record<GridRowId, Record<GridColDef['field'], number>>;
+  hiddenCells: Record<GridRowId, Record<GridColDef['field'], boolean>>;
+  /**
+   * For each hidden cell, it contains the row index corresponding to the cell that is
+   * the origin of the hidden cell. i.e. the cell which is spanned.
+   * Used by the virtualization to properly keep the spanned cells in view.
+   */
+  hiddenCellOriginMap: Record<number, Record<GridColDef['field'], number>>;
+}
+
+const EMPTY_STATE = { spannedCells: {}, hiddenCells: {}, hiddenCellOriginMap: {} };
+const EMPTY_RANGE = { firstRowIndex: 0, lastRowIndex: 0 };
+const skippedFields = new Set(['__check__', '__reorder__']);
+
+function isUninitializedRowContext(renderContext: GridRenderContext) {
+  return renderContext.firstRowIndex === 0 && renderContext.lastRowIndex === 0;
+}
+
+function isRowRenderContextUpdated(
+  prevRenderContext: GridRenderContext,
+  renderContext: GridRenderContext,
+) {
+  return (
+    prevRenderContext.firstRowIndex !== renderContext.firstRowIndex ||
+    prevRenderContext.lastRowIndex !== renderContext.lastRowIndex
+  );
+}
+
+function getUnprocessedRange(
+  testRange: { firstRowIndex: number; lastRowIndex: number },
+  processedRange: { firstRowIndex: number; lastRowIndex: number },
+) {
+  if (
+    testRange.firstRowIndex >= processedRange.firstRowIndex &&
+    testRange.lastRowIndex <= processedRange.lastRowIndex
+  ) {
+    return null;
+  }
+  // Overflowing at the end
+  // Example: testRange={ firstRowIndex: 10, lastRowIndex: 20 }, processedRange={ firstRowIndex: 0, lastRowIndex: 15 }
+  // Unprocessed Range={ firstRowIndex: 16, lastRowIndex: 20 }
+  if (
+    testRange.firstRowIndex >= processedRange.firstRowIndex &&
+    testRange.lastRowIndex > processedRange.lastRowIndex
+  ) {
+    return { firstRowIndex: processedRange.lastRowIndex, lastRowIndex: testRange.lastRowIndex };
+  }
+  // Overflowing at the beginning
+  // Example: testRange={ firstRowIndex: 0, lastRowIndex: 20 }, processedRange={ firstRowIndex: 16, lastRowIndex: 30 }
+  // Unprocessed Range={ firstRowIndex: 0, lastRowIndex: 15 }
+  if (
+    testRange.firstRowIndex < processedRange.firstRowIndex &&
+    testRange.lastRowIndex <= processedRange.lastRowIndex
+  ) {
+    return {
+      firstRowIndex: testRange.firstRowIndex,
+      lastRowIndex: processedRange.firstRowIndex - 1,
+    };
+  }
+  // TODO: Should return two ranges handle overflowing at both ends ?
+  return testRange;
+}
+
+export const rowSpanningStateInitializer: GridStateInitializer = (state) => {
+  return {
+    ...state,
+    rowSpanning: EMPTY_STATE,
```

Comment:
Do you think having a better initial value here is possible to avoid this flickering?

https://github.com/user-attachments/assets/e988f670-0cc2-485e-ae90-58cbb6491760

I know that the `renderContext` is empty at this point, but perhaps we can assume that calculating row span for the first ~20 rows would be enough?

Reply:
Nice idea. Implemented in this [commit](https://github.com/mui/mui-x/pull/14124/commits/cc349114a8fd67360d90b9a9fd654e07cdfc7586).
Is 20 an optimal number though? For Grids with an initial viewport exceeding 20 rows, flickering would still be added after 20 rows.

Reply:
Looks great!

> Is 20 an optimal number though?

We don't know the dimensions of the grid yet, but we can derive the number of rows from 3 things:
- viewport height
- `rowHeight` prop
- density

This should be enough to ensure no flickering in the viewport. What do you think?

Reply:
I think guessing the viewport height at the initialization phase would not be possible, as no ref has been attached to a DOM element yet.

Is there another way that I'm not aware of?

Reply:
Yes, I meant the screen viewport.
We don't know the height of the data grid, so assuming the screen height would guarantee no flickering on the initial render. What do you think?

Reply:
Ah, I got it now.

I suspect it may cause a significant slowdown in the initial render on the high-resolution devices. Imagine a viewport that can fit in 500 rows and the grid has 100 columns, it'll be at least 10,000 comparisons on the initial render.
I'm inclined to go with a fixed number for now, 20 seems a good starting point. It should cover more than 80% of the use cases.

Wdyt?

Reply:
I'm fine with 20 as an initial value 👍🏻